fetcher/fetcher.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fetcher
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+
+	"github.com/ava-labs/hypersdk/keys"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+// Fetcher retrieves values on-the-fly and ensures that
+// a value is only fetched from disk once. Subsequent
+// requests can be retrieved from cache.
+type Fetcher struct {
+	im        state.Immutable
+	cacheLock sync.RWMutex
+	cache     map[string]*fetchData
+
+	keysToFetch map[string][]ids.ID        // Number of txns waiting for a key
+	txnsToFetch map[ids.ID]*sync.WaitGroup // Number of keys a txn is waiting on
+	keyLock     sync.RWMutex
+	txnLock     sync.Mutex
+
+	stopOnce  sync.Once
+	stop      chan struct{}
+	err       error
+	fetchable chan *task
+	l         sync.Mutex
+	wg        sync.WaitGroup
+
+	completed int
+	numTxs    int
+}
+
+// Data to insert into the cache
+type fetchData struct {
+	v      []byte
+	exists bool
+	chunks uint16
+}
+
+// Holds the information that a worker needs to fetch values
+type task struct {
+	ctx context.Context
+	id  ids.ID
+	key string
+	wg  *sync.WaitGroup
+}
+
+// New creates a new [Fetcher]
+func New(numTxs int, concurrency int, im state.Immutable) *Fetcher {
+	f := &Fetcher{
+		im:          im,
+		cache:       make(map[string]*fetchData, numTxs),
+		keysToFetch: make(map[string][]ids.ID),
+		txnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task),
+		stop:        make(chan struct{}),
+		numTxs:      numTxs,
+	}
+	for i := 0; i < concurrency; i++ {
+		f.wg.Add(1)
+		go f.runWorker()
+	}
+	return f
+}
+
+// Workers fetch individual keys
+func (f *Fetcher) runWorker() {
+	defer f.wg.Done()
+	for {
+		select {
+		case t, ok := <-f.fetchable:
+			if !ok {
+				return
+			}
+
+			// Checks the cache twice. In isFirst, we will wait if it's not
+			// in the cache and not the first tx requesting the key. We have
+			// the second check to ensure when the tx is unblocked, it can
+			// fetch from cache.
+			f.isFirst(t)
```

Comment:
I'm thinking about it a bit more and I don't think we need to check the cache twice? Either it's not in the cache AND not the first tx requesting the key so we will block. When we get unblocked, we know that we're only getting unblocked because all of our keys were fetched so we can just `continue`.

However, if it's in the cache then we just proceed as usual and decrement the counts using `f.updateDependencies`. This saves an extra call to `f.has(t.key)`